### PR TITLE
Feature: Added support for reordering items in the sidebar

### LINKED ISSUE #11199: Feature: Add back ability to reorder pinned items
### What feature or improvement do you think would benefit Files?

Before syncing the pinned items with Windows File Explorer, the app used to allow for moving the individual pins around.
Adding back that functionality would remove the need to go into the File Explorer App every time that you want to move a pin.
This is especially useful for newly added pins that appear at the bottom by default.

This functionality could either be in the form of Drag & Drop or an option inside the context menu of a pinned item.

### Requirements

- Add an option in the pinned item context menu to move the item up / down

or

- Add the option to Drag & Drop the pinned items in the list

### Files Version

2.4.31.0

### Windows Version

Windows 10 22H2 19045.2546

### Comments

_No response_

### PR DESCRIPTION
**Pull request by hecksmosis**
**Resolved / Related Issues**
Items resolved / related issues by this PR.
- Closes #11199 

**Validation**
How did you test these changes?
- [x] Built and ran the app
- [x] Tested the changes for accessibility

**Screenshots (optional)**
Add screenshots here.

### CODE REVIEW

**Comment by hecksmosis** [comment]
The Glyph currently used for the icon in the items needs to be changed, does anyone know of a better icon to use there?

**Comment by yaira2** [comment]
Which icon?

**Review by ferrariofilippo — CHANGES_REQUESTED** [review]
Everything seems to work in debug mode, but there are two problems:
- If I run Files without debug mode, half the time I use this feature, the app crashes. Nothing useful is logged.
- We are wasting time (and space) writing the same things in the logs twice

**Comment by hecksmosis** [comment]
> Everything seems to work in debug mode, but there are two problems:
> 
> * If I run Files without debug mode, half the time I use this feature, the app crashes. Nothing useful is logged.
> * We are wasting time (and space) writing the same things in the logs twice
>   

Yes, I forgot to remove the logs, they are totally unnecessary

**Comment by hecksmosis** [comment]
Do you have any recent folders in the quick access widget?

**Comment by hecksmosis** [comment]
By the way, I can't reproduce the crashes on my end, any idea why they could be happening?

**Comment by yaira2** [comment]
Two things I noticed
1. Clicking a pinned item often triggers the drag event even when it was just a regular click
2. It's hard to drag and drop to the right area without accidentally moving the folder into one of the other pinned items. File Explorer has a visual indicator to show where the item is being dropped but I'm not sure we can do that with the sidebar.

**Comment by drasticactions** [comment]
Since you at-replied me this ended up in my email. I looked at your PR and I'm not sure you used my workaround, which may speak to @yair100's issue.

For https://github.com/microsoft/microsoft-ui-xaml/issues/3290, I attach the "DragStarting" event to the underling "NavigationViewItemPresenter". That's a "workaround" (Hack) that lets me attach the same normal events that would be allowed for Elements to the NavigationItem. You can see it implemented here,https://github.com/drasticactions/MauiFeed/blob/main/src/MauiFeed.WinUI/Views/FeedSidebarItem.cs#L237-L252

In short, you would implement it the same way you would have to NavigationViewItem, but instead attach those events to the NavigationViewItemPresenter. Eventually, when the underling issue is fixed, you should be able to add those events back to the NavigationViewItem.

**Comment by hecksmosis** [comment]
> Since you at-replied me this ended up in my email. I looked at your PR and I'm not sure you used my workaround, which may speak to @yair100's issue.
> 
> For [microsoft/microsoft-ui-xaml#3290](https://github.com/microsoft/microsoft-ui-xaml/issues/3290), I attach the "DragStarting" event to the underling "NavigationViewItemPresenter". That's a "workaround" (Hack) that lets me attach the same normal events that would be allowed for Elements to the NavigationItem. You can see it implemented here,https://github.com/drasticactions/MauiFeed/blob/main/src/MauiFeed.WinUI/Views/FeedSidebarItem.cs#L237-L252
> 
> In short, you would implement it the same way you would have to NavigationViewItem, but instead attach those events to the NavigationViewItemPresenter. Eventually, when the underling issue is fixed, you should be able to add those events back to the NavigationViewItem.

True, I ended up not really using your workaround (sorry for the ping)

**Comment by hecksmosis** [comment]
I added the dialog but I am trash at UI so I need help with the buttons to reorder the items in the dialog's navigationview

**Comment by yaira2** [comment]
@hecksmosis can you add drag & drop to the dialog and I'll see if I can figure out the issue?

File: src/Files.App/UserControls/SidebarControl.xaml.cs [hunk 1]
```diff
@@ -331,6 +343,16 @@ private void HideSection()
 			}
 		}
 
+		private async void ReorderItems()
+		{
+			var dialog = new ReorderSidebarItemsDialogViewModel();
+			var dialogService = Ioc.Default.GetRequiredService<IDialogService>();
+			App.Logger.Warn("Reorder items dialog popup");
```
Comment (ferrariofilippo) [inline src/Files.App/UserControls/SidebarControl.xaml.cs:350]: Why is it logged as a warning? I think it is supposed to be an `Info` log

File: src/Files.App/UserControls/SidebarControl.xaml.cs [hunk 2]
```diff
@@ -331,6 +343,16 @@ private void HideSection()
...
+			var result = await dialogService.ShowDialogAsync(dialog);
+
+			App.Logger.Warn("Result: " + result.ToString());
```
Comment (ferrariofilippo) [inline src/Files.App/UserControls/SidebarControl.xaml.cs:353]: As above

File: src/Files.App/UserControls/SidebarControl.xaml.cs [hunk 3]
```diff
@@ -687,7 +681,7 @@ private async void NavigationViewLocationItem_Drop(object sender, DragEventArgs
 
 				isDropOnProcess = false;
 				deferral.Complete();
-			}
+			}
```
Comment (ferrariofilippo) [inline src/Files.App/UserControls/SidebarControl.xaml.cs:684]: Can you remove empty space

**Review by ferrariofilippo — CHANGES_REQUESTED** [review]
I left some comments on the code style

File: src/Files.App/UserControls/SidebarControl.xaml.cs (hunk 1 above)
Comment (hecksmosis) [reply src/Files.App/UserControls/SidebarControl.xaml.cs:350]: It was just there for testing, i'll remove it altogether

File: src/Files.App/Files.App.csproj [hunk 4]
```diff
@@ -64,6 +64,7 @@
         <Content Remove="Assets\FilesHome.png" />
     </ItemGroup>
     <ItemGroup>
+      <None Remove="Dialogs\ReorderSidebarItemsDialog.xaml" />
       <None Remove="UserControls\OpacityIcon.xaml" />
```
[the remark this answers is not in the document]
Comment (hecksmosis) [reply src/Files.App/Files.App.csproj:68]: This was added automatically by VS, I'll remove it if it's not necessary

**Comment by hecksmosis** [comment]
@yair100 I added drag and drop again and fixed the issue

File: src/Files.App/Dialogs/ReorderSidebarItemsDialog.xaml [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+	DefaultButton="Primary"
+	IsPrimaryButtonEnabled="True"
+	PrimaryButtonCommand="{x:Bind ViewModel.PrimaryButtonCommand}"
+	PrimaryButtonText="{helpers:ResourceString Name=Save}"
+	RequestedTheme="{x:Bind helpers:ThemeHelper.RootTheme}"
+	SecondaryButtonText="{helpers:ResourceString Name=Cancel}"
+	Style="{StaticResource DefaultContentDialogStyle}"
+	mc:Ignorable="d">
+
+	<Border Width="400">
```
Comment (yaira2) [inline src/Files.App/Dialogs/ReorderSidebarItemsDialog.xaml:19]: The border can be removed.

File: src/Files.App/UserControls/SidebarControl.xaml.cs [hunk 6]
```diff
@@ -189,6 +194,13 @@ private List<ContextMenuFlyoutItemViewModel> GetLocationItemMenuItems(INavigatio
 
 			return new List<ContextMenuFlyoutItemViewModel>()
 			{
+				new ContextMenuFlyoutItemViewModel()
+				{
+					Text = "ReorderSidebarItemsDialogText".GetLocalizedResource(),
+					Glyph = "\uE8D8",
+					Command = ReorderItemsCommand,
+					ShowItem = isFavoriteItem || item.Section is SectionType.Favorites
+				},
```
Comment (yaira2) [inline src/Files.App/UserControls/SidebarControl.xaml.cs:203]: Can you move to the end of the list?

Comment (hecksmosis) [reply src/Files.App/UserControls/SidebarControl.xaml.cs:203]: Before properties or after?

Comment (yaira2) [reply src/Files.App/UserControls/SidebarControl.xaml.cs:203]: Maybe next to "unpin".

**Comment by hecksmosis** [comment]
What icon should we use for the context menu item?

**Comment by yaira2** [comment]
Let's use the one for `Sort` and if something better comes up we can switch.

**Comment by ferrariofilippo** [comment]
> It seems that `Favorites section` is refreshed twice when you pin/unpin an item ~(I don't know if this behaviour is related with this PR)~

I can't reproduce this problem in the Preview version so it's related to this PR

**Comment by hecksmosis** [comment]
Done

**Comment by ferrariofilippo** [comment]
> I can't reproduce this problem in the Preview version so it's related to this PR

This has been fixed but now I have to restart Files to see the pinned items reordered.
It's wonderful how fixing a bug creates another

**Comment by hecksmosis** [comment]
Yay, I'm blind

**Comment by hecksmosis** [comment]
Should be done now (hopefully no more bugs)

File: src/Files.App/Dialogs/ReorderSidebarItemsDialog.xaml [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+			<ColumnDefinition />
+			<ColumnDefinition Width="Auto" />
+		</Grid.ColumnDefinitions>
+		<Grid.RowDefinitions>
+			<RowDefinition Height="Auto" />
+			<RowDefinition Height="Auto" />
+			<RowDefinition Height="Auto" />
+		</Grid.RowDefinitions>
+
+		<NavigationView
```
Comment (yaira2) [inline src/Files.App/Dialogs/ReorderSidebarItemsDialog.xaml:33]: NavigationView is a pretty heavy control, can you see if ListView works instead?

File: src/Files.App/Files.App.csproj (hunk 4 above)
[the remark this answers is not in the document]
Comment (yaira2) [reply src/Files.App/Files.App.csproj:72]: @onein528 should we leave this in or take it out?

[the remark this answers is not in the document]
Comment (onein528) [reply src/Files.App/Files.App.csproj:72]: YES

File: src/Files.App/Files.App.csproj [hunk 8]
```diff
@@ -121,6 +122,9 @@
         <TrimmerRootAssembly Include="CommunityToolkit.WinUI.UI.Controls.Primitives" />
     </ItemGroup>
     <ItemGroup>
+      <Page Update="Dialogs\ReorderSidebarItemsDialog.xaml">
+        <XamlRuntime>$(DefaultXamlRuntime)</XamlRuntime>
+      </Page>
```
[the remark this answers is not in the document]
Comment (onein528) [reply src/Files.App/Files.App.csproj:131]: Please remove it as above.

File: src/Files.App/Dialogs/ReorderSidebarItemsDialog.xaml (hunk 7 above)
Comment (hecksmosis) [reply src/Files.App/Dialogs/ReorderSidebarItemsDialog.xaml:33]: List view works, but I couldn't find a way to use the icon without using NavigationViewItems as children (and they caused problems with the selection box size iirc)

**Comment by ferrariofilippo** [comment]
> Should be done now (hopefully no more bugs)

Everything is fine now 👍

File: src/Files.App/Dialogs/ReorderSidebarItemsDialog.xaml (hunk 7 above)
Comment (hecksmosis) [reply src/Files.App/Dialogs/ReorderSidebarItemsDialog.xaml:33]: I can't get it to work with listview, the DataContext is breaking.

Comment (yaira2) [reply src/Files.App/Dialogs/ReorderSidebarItemsDialog.xaml:33]: We need a new DataContext.

Comment (hecksmosis) [reply src/Files.App/Dialogs/ReorderSidebarItemsDialog.xaml:33]: How would that be done?

Comment (yaira2) [reply src/Files.App/Dialogs/ReorderSidebarItemsDialog.xaml:33]: What do you mean by DataContext is breaking?

Comment (hecksmosis) [reply src/Files.App/Dialogs/ReorderSidebarItemsDialog.xaml:33]: It throws a StackOverflow when the dialog is created

Comment (yaira2) [reply src/Files.App/Dialogs/ReorderSidebarItemsDialog.xaml:33]: Can you push the commit anyways? I'll see if I can reproduce and debug the issue.

Comment (hecksmosis) [reply src/Files.App/Dialogs/ReorderSidebarItemsDialog.xaml:33]: Ok

Comment (hecksmosis) [reply src/Files.App/Dialogs/ReorderSidebarItemsDialog.xaml:33]: It's probably me being stupid but I can't figure it out

File: src/Files.App/Dialogs/ReorderSidebarItemsDialog.xaml.cs [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+using System;
+using System.Diagnostics;
+using System.Threading.Tasks;
+using Windows.ApplicationModel.DataTransfer;
+
+namespace Files.App.Dialogs
+{
+	public sealed partial class ReorderSidebarItemsDialog : ContentDialog, IDialog<ReorderSidebarItemsDialogViewModel>
+	{
+		public ReorderSidebarItemsDialogViewModel ViewModel
```
Comment (yaira2) [inline src/Files.App/Dialogs/ReorderSidebarItemsDialog.xaml.cs:21]: Where is this ViewModel?

Comment (hecksmosis) [reply src/Files.App/Dialogs/ReorderSidebarItemsDialog.xaml.cs:21]: It's here: `src/Files.App/ViewModels/Dialogs/ReorderSidebarItemsDialogViewModel.cs`

**Comment by hecksmosis** [comment]
I changed the ListViewItem to Grid and it now works fine, this is ready now

**Comment by yaira2** [comment]
Should we display a progress ring and delay closing the dialog until the changes are saved?

**Comment by hecksmosis** [comment]
I don't think this is necessary, this is probably not going to be used very regularly, just occasionally

**Comment by yaira2** [comment]
The reason I think it's worth doing this (it's not a blocker) is that the app freezes and it's not exactly clear to the user what's happening in the background.

**Comment by hecksmosis** [comment]
Oh ok

**Review by yair100 — APPROVED** [review]
LGTM

**Comment by hecksmosis** [comment]
Should I open a PR with the progress ring?

**Comment by yaira2** [comment]
If you don't mind, thanks!

**Comment by Wolfarelli** [comment]
The reorder dialog seems to need a scroll bar for favorites lists longer than 15 items; mouse wheel doesn't work neither.
Captura de pantalla 2023-04-10 074545

**Comment by yaira2** [comment]
@Wolfarelli can you open a feature request for this?
